include/openPMD/auxiliary/Mpi.hpp

```diff
@@ -0,0 +1,83 @@
+/* Copyright 2022 Franz Poeschel
+ *
+ * This file is part of openPMD-api.
+ *
+ * openPMD-api is free software: you can redistribute it and/or modify
+ * it under the terms of of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * openPMD-api is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with openPMD-api.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "openPMD/config.hpp"
+
+#include <optional>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+#if openPMD_HAVE_MPI
+#include <mpi.h>
+#endif
+
+namespace openPMD::auxiliary
+{
+#if openPMD_HAVE_MPI
+
+using Mock_MPI_Comm = MPI_Comm;
+
+namespace
+{
+    template <typename>
+    struct MPI_Types;
```

Comment:
Future refactoring: since this is an openPMD internal type trait, we should consider renaming this so it does not collide with the `MPI_....` prefix used as convention in the MPI C libraries for all its functions.

Reply:
Looking at this, I'm asking myself if this code is even correct. If an MPI application runs on a heterogeneous system with differing CPU architectures, there will be undefined behavior, no?
Maybe, we should just use the largest numeric type for those cases? Or is there a better way?

Reply:
This depends a bit and would require more work, but I think not in the location here.

> heterogeneous system with differing CPU architectures

The questions here is not really the CPU (micro) architecture but the data model: https://en.cppreference.com/w/cpp/language/types Essentially all modern systems you would put on an HPC system are `LP64`.

If you would like to run with MPI on a system with heterogeneous hosts, you would have to compile your binaries for each of them and spawn them with MPI. That's the easy part, the hard part would be if someone ever implemented an MPI that would allow to communicate between LP64 and, let's say ILP32, or LP64 with big and little endian.

Either way, it would not change the local code in this location.

Reply:
I think we can actually simplify this by using `static constexpr MPI_Datatype const value = ...;`

Reply:
We can probably even make this a simpler `constexpr` function now, which might also benefit compile time.

Reply:
`constexpr` will probably not work fully since the MPI datatypes are not `constexpr`, but we can still replace the specialized class template with `if constexpr`, yeah.

> If you would like to run with MPI on a system with heterogeneous hosts, you would have to compile your binaries for each of them and spawn them with MPI. That's the easy part, the hard part would be if someone ever implemented an MPI that would allow to communicate between LP64 and, let's say ILP32, or LP64 with big and little endian.

> Either way, it would not change the local code in this location.

Fair enough, thanks for clarifying!